File: solid/excel/src/js-sheet/parser.ts
import { formatCellRef, isCellAddress, normalizeAddr, parseCellRef, shiftCellRef } from './address'
import type { CellRef, Expr } from './types'

type BinOp = Extract<Expr, { kind: 'binop' }>['op']

function precedence(op: BinOp): number {
  switch (op) {
    case '=':
    case '<>':
    case '<':
    case '<=':
    case '>':
    case '>=':
      return 1
    case '&':
      return 2
    case '+':
    case '-':
      return 3
    case '*':
    case '/':
      return 4
    case '^':
      return 5
  }
  return 0
}

function formatNumber(value: number): string {
  return value === Math.floor(value) && Math.abs(value) < 1e15
    ? String(Math.trunc(value))
    : String(value)
}

function serializeExpr(expr: Expr, parentPrecedence = 0): string {
  switch (expr.kind) {
    case 'number':
      return formatNumber(expr.value)
    case 'text':
      return `"${expr.value}"`
    case 'boolean':
      return expr.value ? 'TRUE' : 'FALSE'
    case 'error':
      return expr.value
    case 'cell':
      return formatCellRef(expr.ref)
    case 'range':
      if (expr.start.sheetName && expr.start.sheetName === expr.end.sheetName) {
        const start = formatCellRef(expr.start)
        const end = formatCellRef({ ...expr.end, sheetName: undefined })
        return `${start}:${end}`
      }
      return `${formatCellRef(expr.start)}:${formatCellRef(expr.end)}`
    case 'negate': {
      const inner = serializeExpr(expr.expr, 6)
      return expr.expr.kind === 'binop' ? `-(${inner})` : `-${inner}`
    }
    case 'func':
      return `${expr.name}(${expr.args.map((arg) => serializeExpr(arg)).join(',')})`
    case 'binop': {
      const current = precedence(expr.op)
      const left = serializeExpr(expr.left, current)
      const right = serializeExpr(expr.right, current + (expr.op === '^' ? 1 : 0))
      const rendered = `${left}${expr.op}${right}`
      return current < parentPrecedence ? `(${rendered})` : rendered
    }
  }
}

function shiftExpr(expr: Expr, rowDelta: number, colDelta: number): Expr {
  switch (expr.kind) {
    case 'number':
      return { kind: 'number', value: expr.value }
    case 'text':
      return { kind: 'text', value: expr.value }
    case 'boolean':
      return { kind: 'boolean', value: expr.value }
    case 'error':
      return { kind: 'error', value: expr.value }
    case 'cell':
      return { kind: 'cell', ref: shiftCellRef(expr.ref, rowDelta, colDelta) }
    case 'range':
      return {
        kind: 'range',
        start: shiftCellRef(expr.start, rowDelta, colDelta),
        end: shiftCellRef(expr.end, rowDelta, colDelta),
      }
    case 'negate':
      return { kind: 'negate', expr: shiftExpr(expr.expr, rowDelta, colDelta) }
    case 'func':
      return {
        kind: 'func',
        name: expr.name,
        args: expr.args.map((arg) => shiftExpr(arg, rowDelta, colDelta)),
      }
    case 'binop':
      return {
        kind: 'binop',
        op: expr.op,
        left: shiftExpr(expr.left, rowDelta, colDelta),
        right: shiftExpr(expr.right, rowDelta, colDelta),
      }
  }
}

export class Parser {
  private readonly chars: string[]
  private pos = 0

  constructor(input: string) {
    this.chars = Array.from(input)
  }

  parse(): Expr | null {
    const expr = this.parseExpr()
    if (!expr) return null
    this.skipWhitespace()
    return this.pos === this.chars.length ? expr : null
  }

  private peek(): string | undefined {
    return this.chars[this.pos]
  }

  private advance(): string | undefined {
    const char = this.chars[this.pos]
    this.pos += 1
    return char
  }

  private skipWhitespace() {
    while (this.peek()?.trim() === '') {
      this.pos += 1
    }
  }

  private expect(expected: string): boolean {
    this.skipWhitespace()
    if (this.peek() !== expected) return false
    this.advance()
    return true
  }

  private parseExpr(): Expr | null {
    return this.parseComparison()
  }

  private parseComparison(): Expr | null {
    this.skipWhitespace()
    let left = this.parseConcat()
    if (!left) return null

    while (true) {
      this.skipWhitespace()
      const op = this.readComparisonOperator()
      if (!op) break
      const right = this.parseConcat()
      if (!right) return null
      left = { kind: 'binop', op, left, right }
    }

    return left
  }

  private parseConcat(): Expr | null {
    this.skipWhitespace()
    let left = this.parseAddSub()
    if (!left) return null

    while (true) {
      this.skipWhitespace()
      if (this.peek() !== '&') break
      this.advance()
      const right = this.parseAddSub()
      if (!right) return null
      left = { kind: 'binop', op: '&', left, right }
    }

    return left
  }

  private parseAddSub(): Expr | null {
    this.skipWhitespace()
    let left = this.parseTerm()
    if (!left) return null

    while (true) {
      this.skipWhitespace()
      const next = this.peek()
      if (next !== '+' && next !== '-') break
      this.advance()
      const right = this.parseTerm()
      if (!right) return null
      left = { kind: 'binop', op: next, left, right }
    }

    return left
  }

  private parseTerm(): Expr | null {
    this.skipWhitespace()
    let left = this.parsePower()
    if (!left) return null

    while (true) {
      this.skipWhitespace()
      const next = this.peek()
      if (next !== '*' && next !== '/') break
      this.advance()
      const right = this.parsePower()
      if (!right) return null
      left = { kind: 'binop', op: next, left, right }
    }

    return left
  }

  private parsePower(): Expr | null {
    this.skipWhitespace()
    const left = this.parseUnary()
    if (!left) return null

    this.skipWhitespace()
    if (this.peek() !== '^') return left
    this.advance()

    const right = this.parsePower()
    if (!right) return null
    return { kind: 'binop', op: '^', left, right }
  }

  private parseUnary(): Expr | null {
    this.skipWhitespace()
    if (this.peek() === '-') {
      this.advance()
      const expr = this.parseUnary()
      return expr ? { kind: 'negate', expr } : null
    }
    return this.parsePrimary()
  }

  private parsePrimary(): Expr | null {
    this.skipWhitespace()
    const next = this.peek()
    if (!next) return null

    if (next === '(') {
      this.advance()
      const expr = this.parseExpr()
      if (!expr || !this.expect(')')) return null
      return expr
    }

    if (next === '"') {
      return this.parseString()
    }

    if (next === '#') {
      return this.parseErrorLiteral()
    }

    if ((next >= '0' && next <= '9') || next === '.') {
      return this.parseNumber()
    }

    if ((next >= 'A' && next <= 'Z') || (next >= 'a' && next <= 'z') || next === '$') {
      return this.parseIdentifierOrReference()
    }

    return null
  }

  private parseNumber(): Expr | null {
    const start = this.pos
    while (true) {
      const next = this.peek()
      if (!next || (!((next >= '0' && next <= '9') || next === '.'))) break
      this.advance()
    }

    const raw = this.chars.slice(start, this.pos).join('')
    const value = Number.parseFloat(raw)
    return Number.isNaN(value) ? null : { kind: 'number', value }
  }

  private parseString(): Expr | null {
    this.advance()
    const start = this.pos
    while (this.peek() !== undefined && this.peek() !== '"') {
      this.advance()
    }
    if (this.peek() !== '"') return null

    const value = this.chars.slice(start, this.pos).join('')
    this.advance()
    return { kind: 'text', value }
  }

  private parseErrorLiteral(): Expr | null {
    const candidates = ['#DIV/0!', '#REF!', '#VALUE!', '#NAME?', '#CYCLE!'] as const
    for (const candidate of candidates) {
      const chars = Array.from(candidate)
      const matches = chars.every((char, index) => this.chars[this.pos + index] === char)
      if (matches) {
        this.pos += chars.length
        return { kind: 'error', value: candidate }
      }
    }
    return null
  }

  private parseIdentifierOrReference(): Expr | null {
    const start = this.pos
    while (true) {
      const next = this.peek()
      if (!next) break
      const isAlphaNum =
        ((next >= 'A' && next <= 'Z')
          || (next >= 'a' && next <= 'z')
          || (next >= '0' && next <= '9')
          || next === '$'
          || next === '_')
      if (!isAlphaNum) break
      this.advance()
    }

    const ident = this.chars.slice(start, this.pos).join('')
    this.skipWhitespace()

    if (this.peek() === '(') {
      this.advance()
      const args = this.parseFuncArgs()
      if (!args || !this.expect(')')) return null
      return { kind: 'func', name: ident.toUpperCase(), args }
    }

    const upper = ident.toUpperCase()
    if (upper === 'TRUE' || upper === 'FALSE') {
      return { kind: 'boolean', value: upper === 'TRUE' }
    }

    let referenceToken = ident
    let sheetName: string | undefined
    if (this.peek() === '!') {
      sheetName = ident
      this.advance()
      const refStart = this.pos
      while (true) {
        const next = this.peek()
        if (!next) break
        const isAlphaNum =
          ((next >= 'A' && next <= 'Z')
            || (next >= 'a' && next <= 'z')
            || (next >= '0' && next <= '9')
            || next === '$')
        if (!isAlphaNum) break
        this.advance()
      }
      referenceToken = `${sheetName}!${this.chars.slice(refStart, this.pos).join('')}`
    }

    let reference: CellRef
    try {
      reference = parseCellRef(referenceToken)
    } catch {
      return null
    }

    this.skipWhitespace()
    if (this.peek() === ':') {
      this.advance()
      this.skipWhitespace()
      const rangeStart = this.pos
      while (true) {
        const next = this.peek()
        if (!next) break
        const isAlphaNum =
          ((next >= 'A' && next <= 'Z')
            || (next >= 'a' && next <= 'z')
            || (next >= '0' && next <= '9')
            || next === '$'
            || next === '_'
            || next === '!')
        if (!isAlphaNum) break
        this.advance()
      }

      const endRaw = this.chars.slice(rangeStart, this.pos).join('')
      let end: CellRef
      try {
        end = parseCellRef(endRaw)
      } catch {
        if (!sheetName) return null
        try {
          end = parseCellRef(`${sheetName}!${endRaw}`)
        } catch {
          return null
        }
      }
      return { kind: 'range', start: reference, end }
    }

    return { kind: 'cell', ref: reference }
  }

  private readComparisonOperator(): '=' | '<>' | '<' | '<=' | '>' | '>=' | null {
    const current = this.peek()
    const next = this.chars[this.pos + 1]

    if (current === '<' && next === '=') {
      this.pos += 2
      return '<='
    }
    if (current === '>' && next === '=') {
      this.pos += 2
      return '>='
    }
    if (current === '<' && next === '>') {
      this.pos += 2
      return '<>'
    }
    if (current === '<' || current === '>' || current === '=') {
      this.pos += 1
      return current
    }
    return null
  }

  private parseFuncArgs(): Expr[] | null {
    this.skipWhitespace()
    if (this.peek() === ')') return []

    const args: Expr[] = []
    while (true) {
      const arg = this.parseExpr()
      if (!arg) return null
      args.push(arg)

      this.skipWhitespace()
      if (this.peek() !== ',') break
      this.advance()
    }

    return args
  }
}

export function parseFormula(input: string): Expr | null {
  const trimmed = input.trim()
  if (!trimmed.startsWith('=')) return null
  return new Parser(trimmed.slice(1)).parse()
}

export function serializeFormula(expr: Expr): string {
  return `=${serializeExpr(expr)}`
}

export function shiftFormulaInput(input: string, rowDelta: number, colDelta: number): string | null {
  const parsed = parseFormula(input)
  return parsed ? serializeFormula(shiftExpr(parsed, rowDelta, colDelta)) : null
}

export function isFormulaReference(input: string): boolean {
  return isCellAddress(normalizeAddr(input))
}

File: rust/excel-core/src/formula/parser.rs
use einfach_core::ValueError;

use crate::cell::CellReference;

use super::{BinOperator, Expr};

/// Parse a formula string. Must start with '='.
/// Returns None if parsing fails.
pub fn parse_formula(input: &str) -> Option<Expr> {
    let input = input.trim();
    if !input.starts_with('=') {
        return None;
    }
    let mut parser = Parser::new(&input[1..]);
    let expr = parser.parse_expr()?;
    if parser.pos < parser.chars.len() {
        return None;
    }
    Some(expr)
}

struct Parser {
    chars: Vec<char>,
    pos: usize,
}

impl Parser {
    fn new(input: &str) -> Self {
        Parser {
            chars: input.chars().collect(),
            pos: 0,
        }
    }

    fn peek(&self) -> Option<char> {
        self.chars.get(self.pos).copied()
    }

    fn advance(&mut self) -> Option<char> {
        let c = self.chars.get(self.pos).copied()?;
        self.pos += 1;
        Some(c)
    }

    fn skip_whitespace(&mut self) {
        while let Some(c) = self.peek() {
            if c.is_whitespace() {
                self.pos += 1;
            } else {
                break;
            }
        }
    }

    fn expect(&mut self, expected: char) -> Option<()> {
        self.skip_whitespace();
        if self.peek() == Some(expected) {
            self.advance();
            Some(())
        } else {
            None
        }
    }

    /// expr = term (('+' | '-') term)*
    fn parse_expr(&mut self) -> Option<Expr> {
        self.parse_comparison()
    }

    /// comparison = concat (('=' | '<>' | '<' | '<=' | '>' | '>=') concat)?
    fn parse_comparison(&mut self) -> Option<Expr> {
        self.skip_whitespace();
        let mut left = self.parse_concat()?;

        loop {
            self.skip_whitespace();
            let op = if self.starts_with("<=") {
                self.pos += 2;
                Some(BinOperator::LessOrEqual)
            } else if self.starts_with(">=") {
                self.pos += 2;
                Some(BinOperator::GreaterOrEqual)
            } else if self.starts_with("<>") {
                self.pos += 2;
                Some(BinOperator::NotEqual)
            } else if self.peek() == Some('<') {
                self.advance();
                Some(BinOperator::Less)
            } else if self.peek() == Some('>') {
                self.advance();
                Some(BinOperator::Greater)
            } else if self.peek() == Some('=') {
                self.advance();
                Some(BinOperator::Equal)
            } else {
                None
            };

            let Some(op) = op else {
                break;
            };

            let right = self.parse_concat()?;
            left = Expr::BinOp {
                op,
                left: Box::new(left),
                right: Box::new(right),
            };
        }

        Some(left)
    }

    /// concat = add_sub ('&' add_sub)*
    fn parse_concat(&mut self) -> Option<Expr> {
        self.skip_whitespace();
        let mut left = self.parse_add_sub()?;

        loop {
            self.skip_whitespace();
            if self.peek() != Some('&') {
                break;
            }
            self.advance();
            let right = self.parse_add_sub()?;
            left = Expr::BinOp {
                op: BinOperator::Concat,
                left: Box::new(left),
                right: Box::new(right),
            };
        }

        Some(left)
    }

    /// add_sub = term (('+' | '-') term)*
    fn parse_add_sub(&mut self) -> Option<Expr> {
        self.skip_whitespace();
        let mut left = self.parse_term()?;

        loop {
            self.skip_whitespace();
            match self.peek() {
                Some('+') => {
                    self.advance();
                    let right = self.parse_term()?;
                    left = Expr::BinOp {
                        op: BinOperator::Add,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                Some('-') => {
                    self.advance();
                    let right = self.parse_term()?;
                    left = Expr::BinOp {
                        op: BinOperator::Sub,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Some(left)
    }

    /// term = power (('*' | '/') power)*
    fn parse_term(&mut self) -> Option<Expr> {
        self.skip_whitespace();
        let mut left = self.parse_power()?;

        loop {
            self.skip_whitespace();
            match self.peek() {
                Some('*') => {
                    self.advance();
                    let right = self.parse_power()?;
                    left = Expr::BinOp {
                        op: BinOperator::Mul,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                Some('/') => {
                    self.advance();
                    let right = self.parse_power()?;
                    left = Expr::BinOp {
                        op: BinOperator::Div,
                        left: Box::new(left),
                        right: Box::new(right),
                    };
                }
                _ => break,
            }
        }
        Some(left)
    }

    /// power = unary ('^' power)?
    fn parse_power(&mut self) -> Option<Expr> {
        self.skip_whitespace();
        let left = self.parse_unary()?;

        self.skip_whitespace();
        if self.peek() == Some('^') {
            self.advance();
            let right = self.parse_power()?;
            return Some(Expr::BinOp {
                op: BinOperator::Power,
                left: Box::new(left),
                right: Box::new(right),
            });
        }

        Some(left)
    }

    /// unary = '-' unary | primary
    fn parse_unary(&mut self) -> Option<Expr> {
        self.skip_whitespace();
        if self.peek() == Some('-') {
            self.advance();
            let expr = self.parse_unary()?;
            Some(Expr::Negate(Box::new(expr)))
        } else {
            self.parse_primary()
        }
    }

    /// primary = number | string | func_call | cell_ref_or_range | '(' expr ')'
    fn parse_primary(&mut self) -> Option<Expr> {
        self.skip_whitespace();

        match self.peek()? {
            '(' => {
                self.advance();
                let expr = self.parse_expr()?;
                self.expect(')')?;
                Some(expr)
            }
            '"' => self.parse_string(),
            '#' => self.parse_error_literal(),
            c if c.is_ascii_digit() || c == '.' => self.parse_number(),
            c if c.is_ascii_alphabetic() || c == '$' => self.parse_identifier_or_reference(),
            _ => None,
        }
    }

    fn parse_number(&mut self) -> Option<Expr> {
        let start = self.pos;
        while let Some(c) = self.peek() {
            if c.is_ascii_digit() || c == '.' {
                self.advance();
            } else {
                break;
            }
        }
        let s: String = self.chars[start..self.pos].iter().collect();
        let n: f64 = s.parse().ok()?;
        Some(Expr::Number(n))
    }

    fn parse_string(&mut self) -> Option<Expr> {
        self.advance();
        let start = self.pos;
        while let Some(c) = self.peek() {
            if c == '"' {
                let s: String = self.chars[start..self.pos].iter().collect();
                self.advance();
                return Some(Expr::Text(s));
            }
            self.advance();
        }
        None
    }

    fn parse_error_literal(&mut self) -> Option<Expr> {
        const ERRORS: [(&str, ValueError); 5] = [
            ("#DIV/0!", ValueError::DivisionByZero),
            ("#REF!", ValueError::InvalidRef),
            ("#VALUE!", ValueError::InvalidValue),
            ("#NAME?", ValueError::InvalidName),
            ("#CYCLE!", ValueError::CyclicRef),
        ];

        for (text, error) in ERRORS {
            if self.starts_with(text) {
                self.pos += text.chars().count();
                return Some(Expr::Error(error));
            }
        }

        None
    }

    /// Identifier: could be a function name (followed by '(') or a cell reference.
    fn parse_identifier_or_reference(&mut self) -> Option<Expr> {
        let start = self.pos;
        while let Some(c) = self.peek() {
            if c.is_ascii_alphanumeric() || c == '$' || c == '_' {
                self.advance();
            } else {
                break;
            }
        }
        let ident: String = self.chars[start..self.pos].iter().collect();

        self.skip_whitespace();

        if self.peek() == Some('!') {
            self.advance();
            self.skip_whitespace();
            let ref_start = self.pos;
            while let Some(c) = self.peek() {
                if c.is_ascii_alphanumeric() || c == '$' {
                    self.advance();
                } else {
                    break;
                }
            }
            let ref_ident: String = self.chars[ref_start..self.pos].iter().collect();
            let mut reference = CellReference::parse(&ref_ident)?;
            reference.sheet_name = Some(ident.clone());

            self.skip_whitespace();
            if self.peek() == Some(':') {
                self.advance();
                self.skip_whitespace();
                let range_start = self.pos;
                while let Some(c) = self.peek() {
                    if c.is_ascii_alphanumeric() || c == '$' {
                        self.advance();
                    } else {
                        break;
                    }
                }
                let end_ident: String = self.chars[range_start..self.pos].iter().collect();
                let mut end_reference = CellReference::parse(&end_ident)?;
                end_reference.sheet_name = Some(ident);
                return Some(Expr::Range {
                    start: reference,
                    end: end_reference,
                });
            }

            return Some(Expr::CellRef(reference));
        }

        if self.peek() == Some('(') {
            self.advance();
            let args = self.parse_func_args()?;
            self.expect(')')?;
            return Some(Expr::FuncCall {
                name: ident.to_ascii_uppercase(),
                args,
            });
        }

        match ident.to_ascii_uppercase().as_str() {
            "TRUE" => return Some(Expr::Boolean(true)),
            "FALSE" => return Some(Expr::Boolean(false)),
            _ => {}
        }

        if let Some(reference) = CellReference::parse(&ident) {
            self.skip_whitespace();
            if self.peek() == Some(':') {
                self.advance();
                self.skip_whitespace();
                let range_start = self.pos;
                while let Some(c) = self.peek() {
                    if c.is_ascii_alphanumeric() || c == '$' {
                        self.advance();
                    } else {
                        break;
                    }
                }
                let end_ident: String = self.chars[range_start..self.pos].iter().collect();
                let end_reference = CellReference::parse(&end_ident)?;
                return Some(Expr::Range {
                    start: reference,
                    end: end_reference,
                });
            }
            return Some(Expr::CellRef(reference));
        }

        None
    }

    fn starts_with(&self, pattern: &str) -> bool {
        let pattern_chars: Vec<char> = pattern.chars().collect();
        self.chars[self.pos..].starts_with(&pattern_chars)
    }

    fn parse_func_args(&mut self) -> Option<Vec<Expr>> {
        let mut args = Vec::new();
        self.skip_whitespace();

        if self.peek() == Some(')') {
            return Some(args);
        }

        args.push(self.parse_func_arg()?);

        loop {
            self.skip_whitespace();
            if self.peek() == Some(',') {
                self.advance();
                args.push(self.parse_func_arg()?);
            } else {
                break;
            }
        }
        Some(args)
    }

    fn parse_func_arg(&mut self) -> Option<Expr> {
        self.parse_expr()
    }
}
